### ISSUE TITLE
Allocations with broadcasting field vectors, including `fill!(::FieldVector, ::Number)` and `zero(eltype(::FieldVector))`

### ISSUE THREAD

**Issue by charleskawczynski** [CliMA/ClimaCore.jl#983]
ClimaCore incurs allocations with:

 - [ ] `Y .= 1` where `Y::FieldVector`
 - [ ] `fill!(::FieldVector, ::Number)` and
 - [ ] `Y .= zero(eltype(Y))` where `Y::FieldVector`

PR with broken tests are pending.

**Comment by charleskawczynski** [comment]
Also, I just realized, #949 was closed because #950 accidentally tested `Field`s and not `FieldVector`s.

**Comment by charleskawczynski** [comment]
Reproducer:

```julia
# These tests require running with `--check-bounds=[auto|no]`
using Test, StaticArrays, IntervalSets
import ClimaCore: ClimaCore, Fields, slab, Domains, Topologies, Meshes, Operators, Spaces, Geometry
function FieldFromNamedTuple(space, nt::NamedTuple)
    cmv(z) = nt
    return cmv.(Fields.coordinate_field(space))
end
include(joinpath(pkgdir(ClimaCore), "test", "Fields", "util_spaces.jl"))
space = first(all_spaces(Float64));
Y = Fields.FieldVector(; c = FieldFromNamedTuple(space, (; x = Float64(0))),
                         f = FieldFromNamedTuple(space, (; x = Float64(0))));
function bc_val!(Y)
    Y .= 1
    nothing
end
bc_val!(Y) # compile first

using Profile
Profile.Allocs.@profile sample_rate=1 bc_val!(Y)
import PProf
PProf.Allocs.pprof(from_c = false)

using Cthulhu
@descend bc_val!(Y)

using JET
bc_val!(Y)
@test_opt bc_val!(Y)
```
